### PR TITLE
fix: Use pubspec override instead of direct path.

### LINKED ISSUE #1768: Enable pub.dev to validate the pubspec.yaml.
Serverpods pub.dev score is reduced because it fails to verify our repository URL: https://pub.dev/packages/serverpod/score

The reason is that we still have `published_to: none` in our checked in repository that is required to reference packages using relative paths.

In order to allow us to have local dependencies in the project we should instead use the [pubspec_override](https://dart.dev/tools/pub/dependencies#dependency-overrides) to provide a relative path that is ignored by pub.dev.

### PR DESCRIPTION
**Pull request by SandPod**
### Change:
- Utilizes dependency_overrides to reference packages locally when in developer mode.
- Removes all dependency override when building packages for publish and production.

Since the publish_to is removed from the serverpod_cli pubspec we should no longer not get a penalty on pubdev once this is merged. There it should resolve #1768 

As part of this change I had to add a version to serverpod packages for all pub specs that needed local references.

Because the overrides are different compared to just path references I had to add override templates for the examples and the serverpod_test_project to ensure that the dependencies could be resolved correctly.

### Create project in development mode
Screenshot 2024-01-23 at 18 38 59

### Create project in production mode
This required some manual changes of the CLI so that the production templates could be loaded locally.
Screenshot 2024-01-23 at 18 44 34

## Pre-launch Checklist

- [ ] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [ ] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [ ] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [ ] I added new tests to check the change I am making.
- [ ] All existing and new tests are passing.
- [ ] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

_If you have done any breaking changes, make sure to outline them here, so that they can be included in the notes for the next release._

### CODE REVIEW

File: templates/pubspecs/examples/auth_example/auth_example_flutter/pubspec.yaml [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  # The following adds the Cupertino Icons font to your application.
+  # Use with the CupertinoIcons class for iOS style icons.
+  cupertino_icons: ^1.0.5
+
+dev_dependencies:
+  flutter_lints: ^3.0.0
+  flutter_test:
+    sdk: flutter
+
+dependency_overrides: #--CONDITIONALLY_REMOVE_LINE--#
```
Comment (vlidholt) [inline templates/pubspecs/examples/auth_example/auth_example_flutter/pubspec.yaml:47]: Isn't the point of the overrides that we do not need to remove them when publishing?

> If you publish a package to pub.dev, keep in mind that your package’s dependency overrides are ignored by all users of your package.

**Review by vlidholt — CHANGES_REQUESTED** [review]
Great work! I see this as a first step to remove the pubspec templates. It seems safe to keep the `dependency_overrides` in the published files. I.e., we won't need the `#--CONDITIONALLY_REMOVE_LINE--#` in the templates.

A way to test that we are resolving everything correctly locally would be to change the contents of the `VERSION` file to something not published to Pub, e.g., `2.0.0`. Then run, `util/update_pubspecs` and `util/pub_get_all`.

In that case, the only thing left in the templates is `VERSION`, correct? Maybe we can find a way to update the version of all packages simultaneously. It's possible [Melos](https://pub.dev/packages/melos) can do this, or if we write some script of our own. That way, we could get rid of the pubspec templates once and for all.

For this PR, let's just add the overrides though! :)

File: templates/pubspecs/examples/auth_example/auth_example_flutter/pubspec.yaml (hunk 1 above)
Comment (SandPod) [reply templates/pubspecs/examples/auth_example/auth_example_flutter/pubspec.yaml:47]: You are correct, the only places that we really need it are the project creation files where we should always reference the published version (in production mode).

I'll slim down the removal of the dependency override to only places that are necessary.

But I'm not 100% sure what happens inside the published package. Say for example our CLI tools that is published with dependency overrides but is also bundled as a binary, I'm not sure that the binary would ignore the dependency_overrrides, so might be safer to at least strip them there.

**Comment by SandPod** [comment]
@vlidholt Review comments should be adressed. I removed all removals of dependency overrides that are not necessary but kept the ones in the CLI with the motivation above.

Also found 2 more issues of dependency problems using the test you suggested with version 2. We can now resolve the full project even if we use a next major version.

I think we should test this with a pre-release once this is merged to ensure that all dependencies are resolved as they should be.

**Review by vlidholt — APPROVED** [review]
LGTM

**Pull request merged**